hail/src/main/scala/is/hail/expr/types/virtual/TNDArray.scala

```diff
@@ -7,6 +7,17 @@ import org.apache.spark.sql.Row
 
 import scala.reflect.{ClassTag, classTag}
 
+object TNDArray {
+  def matMulNDims(l: Int, r: Int): Int = {
+    (l, r) match {
+      case (1, 1) => 0
+      case (1, n) => n - 1
+      case (n, 1) => n - 1
+      case (_, _) => l
```

Comment:
it's unfortunate that this function is defined in like four places.

Reply:
I can extract it into some misc helper in python so it's duplicated one less time there, but still unfortunate.